src/main/scala/is/hail/expr/ir/functions/Functions.scala

```diff
@@ -0,0 +1,148 @@
+package is.hail.expr.ir.functions
+
+import is.hail.asm4s._
+import is.hail.expr.{AST, FunType, Lambda, TypeTag}
+import is.hail.expr.ir._
+import is.hail.expr.types._
+import is.hail.utils._
+import is.hail.asm4s.coerce
+
+import scala.collection.mutable
+import scala.reflect.ClassTag
+
+object IRFunctionRegistry {
+
+  val registry: mutable.Map[String, Seq[(TypeTag, Seq[IR] => IR)]] = mutable.Map().withDefaultValue(Seq.empty)
+
+  def addIRFunction(f: IRFunction) {
+    val l = registry(f.name)
+    registry.put(f.name,
+      l :+ (f.types, { args: Seq[IR] =>
+        ApplyFunction(f, args)
+      }))
+  }
+
+  def addIR(name: String, types: Seq[Type], f: Seq[IR] => IR) {
+    val l = registry(name)
+    registry.put(name, l :+ ((FunType(types: _*), f)))
+  }
+
+  def lookupFunction(name: String, args: TypeTag): Option[Seq[IR] => IR] = {
+    val validMethods = registry(name).flatMap { case (tt, f) =>
+      if (tt.xs.length == args.xs.length) {
+        tt.clear()
+        if (tt.unify(args))
+          Some(f)
+        else
+          None
+      } else {
+        None
+      }
+    }
+    validMethods match {
+      case Seq() => None
+      case Seq(x) => Some(x)
+      case _ => fatal(s"Multiple methods found that satisfy $name$args.")
+    }
+  }
+
+  UtilFunctions.registerAll()
+}
+
+abstract class RegistryFunctions {
+
+  def registerAll(): Unit
+
+  private val tvs = mutable.Map[String, TVariable]()
+  private val tnums = mutable.Map[String, TVariable]()
+
+  def tv(name: String): TVariable =
+    tvs.getOrElseUpdate(name, TVariable(name))
+
+  def tnum(name: String): TVariable =
+    tnums.getOrElseUpdate(name, TVariable(name, _.isInstanceOf[TNumeric]))
+
+  def tvar(cond: Type => Boolean) = TVariable("COND", cond)
```

Comment:
I think this is starting to break down.  It is surprising that rv("X") and tnum("X") aren't unified.  And why tv and tvar?

Reply:
I could get rid of `tvar`. I was just trying to find a neater way to deal with the `TTuple` case and assuming that there would be other functions that would want to do similar things, but I can just get rid of it and write it in explicitly there.

I guess I don't see why tv("X") and tnum("X") should be unified? If they're supposed to represent the same thing, then they're either both numeric or both not, and you should use one or the other. I could do the thing that we do in the function registry where we just hard-code three TVariables that we can use in function definitions. I left it like this for flexibility, so that in any given function definition you could use whatever names you wanted as long as you were consistent within the definition.

Reply:
Also, I'm considering getting rid of the return type in `registerIR` definitions. I originally had it in for consistency between the various `registerX` methods, but it's not used anywhere (since the IR does its own type inference), and some of the IR functions really don't want to define it.

Reply:
I feel that way because they have the same name and they're in the same scope.  For example, I write:

regsiter("add", tnum("X"), tv("X"), ...)

It seems like I'm saying "make sure the first type is numeric, and call it X, and make sure the second type is the same as the first one."

I like the trick for TTuple, I just think the interface needs work.  Also, why doesn't tvar get a name?